### ISSUE TITLE
Filter by question before plotting/showing counts

### ISSUE THREAD

**Issue by lheckmann** [N2-Survey/SurveyFramework#84]
Introduce a filter option to only select parts of the dataframe when using other functionalities (e.g. plotting).

E.g. Only Germans should be selected for certain questions or only everyone that is not German,...

**Comment by lheckmann** [comment]
I think we do not really need a function for this, but a explanation should do. 
What is more important is that this germans would then not be the input for the .plot function right? So germans.plot('some other question') would not be possible, but is something that will rather be needed maybe?

So the possibilities I see, but we should discuss them together with @rguliev, are:
1. Either make a function that gives you a filtered survey object where you can then use all functionalities again
2. Introduce a filter option (maybe a map as input parameter) to the plot function.

**Comment by rguliev** [comment]
Hi,

I think we can consider:
1) Slicing object itself, i.e. `germans = survey[survey.responses["Nationality"] == "german"]`
2) Add `filter` method, i.e. `germans = survey.filter(survey.responses["Nationality"] == "german")`
3) Or add both options

In both cases, the output is a `LimeSurvey` object as well, so it can be followed with other methods, for example `survey[survey.responses["Nationality"] == "german"].plot("A1")`

**Comment by fudw** [comment]
> In both cases, the output is a `LimeSurvey` object as well, so it can be followed with other methods, for example `survey[survey.responses["Nationality"] == "german"].plot("A1")`

@rguliev If I understood you correctly, a _copy_ of the entire LimeSurvey object is created by such a filter or slicing operation, instead of a _view_ of the survey.responses dataframe. Because otherwise one would not be able to directly call further plotting methods on `germans` here, right?

I was under the impression this is undesirable due to memory consumption of having multiple copies of (part of) the raw response entries df. Do you think this could become a problem if a copy is created every time such a filter is called by the user, which I imagine would be done many times?

**Comment by rguliev** [comment]
That is a good point. What was your plan for this issue then?

If we do not have better options, then I guess will can keep it for now. Also, 
1) I think that the garbage collector will take care of "inactive" filtered objects
2) We optimized the size of the responses, so the copies will be relatively light
3) `survey[survey.responses["Nationality"] == "german", [<some columns>]]` would also allow to reduce the number of columns used

**Comment by fudw** [comment]
I was thinking about incorporating this functionality into the `count` method, where filtering conditions are inputted as optional arguments and the slicing of the responses df (creating a view rather than a copy) is done right before the calculating sums. This way the filtered responses don't need to be saved but are retrieved every time it is called within the plotting functions.

**Comment by fudw** [comment]
For example, if one wants to plot question A5 but only for those who answered "german" to question A3 (I'm making up some IDs here), one would 

1. call `survey.plot("A5", filters=[("A3", "=='german'")])` The `filters` argument would perhaps take a list of tuples as conditions for filtering in the form of `("question_id", "condition")` 
2. This argument will be passed onto the `count` method, which applies the filters to narrow down the `responses` df before calculating the counts and percentages
3. Then the counts based on filtered data will be given to the plotting functions as usual. 

So basically the filtering and slicing is done before the counts are calculated, and creates only views of the presumably large `responses` df rather than copies. The fact that it is handled within `count` is also helpful when one simply wants to look at the counts and percentages themselves, as shown in the `readme.md` usage example section.

**Comment by rguliev** [comment]
I see. We could use `query` parameter and pass it to [pandas query](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html). However, I would suggest not making filters as function arguments. I feel like it's going to be a headache and a bad user experience.

**Comment by rguliev** [comment]
Discussed on the meeting

```python
def filter(filter) -> LimeSurvey:
   tmp = self.copy()
   tmp.resonses = tmp.responses.iloc[filter,:]
   return tmp


def query(filter:str) -> LimeSurvey:
   tmp = self.copy()
   tmp.resonses = tmp.responses.query(filter)
   return tmp


def keep_choices(question:str, choices:list) -> LimeSurvey:
   # self.query("A3 in ('german', ...)")
   return self.query(f"{question} in ({', '.join(choices)})")



s.filter(s.responses.A3=='german')
s.query('A3 == "german"')
s.keep_choices("A3", ["german"]).keep_choices("A1", ["A1"])
```
